In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats import f_oneway
import warnings
warnings.filterwarnings('ignore')

# Data Frame

In [2]:
accident = pd.read_csv('DATASETS\\accident_data.csv')

In [3]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [4]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mode()[0])
accident['Road_Type'] = accident['Road_Type'].fillna(accident['Road_Type'].mode()[0])
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna(accident['Weather_Conditions'].mode()[0])
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])
                                                                                                      

In [5]:
accident['Accident Date'] = accident['Accident Date'].astype('str')
accident['Accident Date'] = accident['Accident Date'].str.strip()
accident['Accident Date'] = accident['Accident Date'].str.replace('/', '-')

# Convert Date 

In [6]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

In [7]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['DayOfWeek'] = accident['Accident Date'].dt.dayofweek

In [8]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayOfWeek                  0
dtype: int64

In [9]:
accident.dtypes

Index                              object
Accident_Severity                  object
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
Year                                int32
Month                               int32
Day                                 int32
DayOfWeek                           int32
dtype: object

In [10]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')

In [11]:
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')

In [12]:
accident['District Area'] = accident['District Area'].astype('category')

In [13]:
accident['Number_of_Casualties'] = accident['Number_of_Casualties'].astype('category')

In [14]:
accident['Number_of_Vehicles'] = accident['Number_of_Vehicles'].astype('category')

In [15]:
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')

In [16]:
accident['Road_Type'] = accident['Road_Type'].astype('category')

In [17]:
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')

In [18]:
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')


In [19]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')

In [20]:
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [21]:
accident['Latitude'] = accident['Latitude'].astype('category')

In [22]:
accident['Longitude'] = accident['Longitude'].astype('category')

In [23]:
accident['Year'] = accident['Year'].astype('category')


In [24]:
accident['Month'] = accident['Month'].astype('category')

In [25]:
accident['Day'] = accident['Day'].astype('category')

In [26]:
accident['DayOfWeek'] = accident['DayOfWeek'].astype('category')

In [27]:
accident.dtypes


Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                         category
Light_Conditions                 category
District Area                    category
Longitude                        category
Number_of_Casualties             category
Number_of_Vehicles               category
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
Year                             category
Month                            category
Day                              category
DayOfWeek                        category
dtype: object

# #1 What date has a more accident rate ?

In [28]:
accident['Accident Date'].mode()[0]


Timestamp('2019-11-30 00:00:00')

# INSIGTH#1

# Most accident happens on 2021-02-11


# #2 What district is commonly involve in accidents?

In [29]:
district_accident = accident['District Area'].mode()[0]
district_accident

'Birmingham'

# INSIGTH#2 

# So most accidents happen on the 'Birmingham' district

# #3 What common location has more accident probability?

In [30]:
accident['Urban_or_Rural_Area'].mode()[0]

'Urban'

# INSIGTH #3 
# So most accident happens on Urban 

In [31]:
df_2019 = accident[accident['Year'] == 2019]


In [32]:
print(f"total number of accident in 2019 is {df_2019['Light_Conditions'].value_counts().sum()}")
df_2019['Light_Conditions'].value_counts()

total number of accident in 2019 is 182115


Light_Conditions
Daylight                       133042
Darkness - lights lit           35769
Darkness - no lighting          10842
Darkness - lighting unknown      1768
Darkness - lights unlit           694
Name: count, dtype: int64

# INSIGTH# 4
# Contrary to popular belief that most accident happens in darkness or at nigth time, most of accident happens on board daytime. About 73% of total accidents happens during daylight

In [33]:
df_2019['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       148010
Raining no high winds     22835
Other                      4083
Raining + high winds       2959
Fine + high winds          2545
Fog or mist                 857
Snowing no high winds       680
Snowing + high winds        146
Name: count, dtype: int64

# INSIGTH # 5
# Out of 182,115 accidents recorded. 79% occured under fine. This indicates that majority of the accidents happen on 2019 happens not because of the bad weather.

In [34]:
casualty_2019 = df_2019.groupby(['Weather_Conditions', 'Vehicle_Type'])['Number_of_Casualties'].count()
casualty_2019

Weather_Conditions     Vehicle_Type                        
Fine + high winds      Agricultural vehicle                       6
                       Bus or coach (17 or more pass seats)     157
                       Car                                     1817
                       Data missing or out of range               0
                       Goods 7.5 tonnes mgw and over             96
                                                               ... 
Snowing no high winds  Other vehicle                              2
                       Pedal cycle                                0
                       Ridden horse                               0
                       Taxi/Private hire car                     20
                       Van / Goods 3.5 tonnes mgw or under       24
Name: Number_of_Casualties, Length: 128, dtype: int64

In [35]:
df_exam3 = df_2019[df_2019['Weather_Conditions'] == 'Fine no high Winds']

In [36]:
df_exam4 = df_exam3.groupby(['Vehicle_Type','Urban_or_Rural_Area'])['Number_of_Casualties'].size()
df_exam4.unstack()

Urban_or_Rural_Area                    Rural  Unallocated  Urban
Vehicle_Type                                                    
Agricultural vehicle                       0            0      0
Bus or coach (17 or more pass seats)       0            0      0
Car                                        0            0      0
Data missing or out of range               0            0      0
Goods 7.5 tonnes mgw and over              0            0      0
Goods over 3.5t. and under 7.5t            0            0      0
Minibus (8 - 16 passenger seats)           0            0      0
Motorcycle 125cc and under                 0            0      0
Motorcycle 50cc and under                  0            0      0
Motorcycle over 125cc and up to 500cc      0            0      0
Motorcycle over 500cc                      0            0      0
Other vehicle                              0            0      0
Pedal cycle                                0            0      0
Ridden horse                               0            0      0
Taxi/Private hire car                      0            0      0
Van / Goods 3.5 tonnes mgw or under        0            0      0

# INSIGTH # 6 
# On 2019, upon closer examination of the vehicle type accidents, most of the accident was in the vehicle type 'Car', 37,652 accidents or 20% of the total of the accident happens on Rural Area.

# INSIGTH # 7
# About 81% of the accident happened on Rural Areas by a Car were clasiofied as 'Sligth', in terms of severity. And les than 20% are Fatal.

# INSIGTH # 8
# In contrast , 35.8% of all accidents happened on 2019  or almost 64.248 incidents involve Car in Urban Area a notably higher figure. AND About 86.6% accident ended up in Sligth severity


In [37]:
accident.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  object        
 1   Accident_Severity        660679 non-null  category      
 2   Accident Date            660679 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  category      
 4   Light_Conditions         660679 non-null  category      
 5   District Area            660679 non-null  category      
 6   Longitude                660679 non-null  category      
 7   Number_of_Casualties     660679 non-null  category      
 8   Number_of_Vehicles       660679 non-null  category      
 9   Road_Surface_Conditions  660679 non-null  category      
 10  Road_Type                660679 non-null  category      
 11  Urban_or_Rural_Area      660679 non-null  category      
 12  Weather_Conditio

In [38]:
df_exam5 = df_exam3.groupby(['Vehicle_Type','Urban_or_Rural_Area','Accident_Severity'])['Number_of_Casualties'].size()
df_exam5.unstack()

Accident_Severity                                          Fatal  Serious  \
Vehicle_Type                          Urban_or_Rural_Area                   
Agricultural vehicle                  Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Bus or coach (17 or more pass seats)  Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Car                                   Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Data missing or out of range          Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Goods 7.5 tonnes mgw and over         Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Goods over 3.5t. and under 7.5t       Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Minibus (8 - 16 passenger seats)      Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Motorcycle 125cc and under            Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Motorcycle 50cc and under             Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Motorcycle over 125cc and up to 500cc Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Motorcycle over 500cc                 Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Other vehicle                         Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Pedal cycle                           Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Ridden horse                          Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Taxi/Private hire car                 Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   
Van / Goods 3.5 tonnes mgw or under   Rural                    0        0   
                                      Unallocated              0        0   
                                      Urban                    0        0   

Accident_Severity                                          Slight  
Vehicle_Type                          Urban_or_Rural_Area          
Agricultural 

# INSGTH # 9
# Motorcycle accidents in rural areas have the highest severity, while cars account for 64,248 accidents in urban areas, where 86.6% are slight, highlighting the need for targeted safety measures in both areas.

# INSIGTH # 10
# Motorcycle accidents in rural areas are more severe, while 64,248 car accidents in urban areas are mostly slight, suggesting the need for targeted safety measures.

In [39]:
df_exam3['Number_of_Casualties'] = pd.to_numeric(df_exam3['Number_of_Casualties'], errors='coerce')

In [40]:
df_exam6 = df_exam3.groupby(['Vehicle_Type', 'Accident_Severity'])['Number_of_Casualties'].sum()
df_exam6_unstacked = df_exam6.unstack()
print(df_exam6_unstacked)

Accident_Severity                      Fatal  Serious  Slight
Vehicle_Type                                                 
Agricultural vehicle                       0        0       0
Bus or coach (17 or more pass seats)       0        0       0
Car                                        0        0       0
Data missing or out of range               0        0       0
Goods 7.5 tonnes mgw and over              0        0       0
Goods over 3.5t. and under 7.5t            0        0       0
Minibus (8 - 16 passenger seats)           0        0       0
Motorcycle 125cc and under                 0        0       0
Motorcycle 50cc and under                  0        0       0
Motorcycle over 125cc and up to 500cc      0        0       0
Motorcycle over 500cc                      0        0       0
Other vehicle                              0        0       0
Pedal cycle                                0        0       0
Ridden horse                               0        0       0
Taxi/Pri

# INSIGTH # 11
# Motorcycle accidents in rural areas have the highest severity, while cars are involved in 64,248 accidents in urban areas, with 86.6% being slight, emphasizing the need for targeted safety measures in both regions

In [41]:
df_exam5 = df_exam3.groupby(['Vehicle_Type', 'Urban_or_Rural_Area'])['Number_of_Casualties'].mean()
df_exam5.unstack()

Urban_or_Rural_Area                    Rural  Unallocated  Urban
Vehicle_Type                                                    
Agricultural vehicle                     NaN          NaN    NaN
Bus or coach (17 or more pass seats)     NaN          NaN    NaN
Car                                      NaN          NaN    NaN
Data missing or out of range             NaN          NaN    NaN
Goods 7.5 tonnes mgw and over            NaN          NaN    NaN
Goods over 3.5t. and under 7.5t          NaN          NaN    NaN
Minibus (8 - 16 passenger seats)         NaN          NaN    NaN
Motorcycle 125cc and under               NaN          NaN    NaN
Motorcycle 50cc and under                NaN          NaN    NaN
Motorcycle over 125cc and up to 500cc    NaN          NaN    NaN
Motorcycle over 500cc                    NaN          NaN    NaN
Other vehicle                            NaN          NaN    NaN
Pedal cycle                              NaN          NaN    NaN
Ridden horse                             NaN          NaN    NaN
Taxi/Private hire car                    NaN          NaN    NaN
Van / Goods 3.5 tonnes mgw or under      NaN          NaN    NaN

# INSIGTH # 12 

# Most  Casualties occure on Motorcycle 

In [42]:
df_exam5 = df_exam3.groupby(['Vehicle_Type', 'Accident_Severity'])['Number_of_Casualties'].median()
df_exam5.unstack()

Accident_Severity                      Fatal  Serious  Slight
Vehicle_Type                                                 
Agricultural vehicle                     NaN      NaN     NaN
Bus or coach (17 or more pass seats)     NaN      NaN     NaN
Car                                      NaN      NaN     NaN
Data missing or out of range             NaN      NaN     NaN
Goods 7.5 tonnes mgw and over            NaN      NaN     NaN
Goods over 3.5t. and under 7.5t          NaN      NaN     NaN
Minibus (8 - 16 passenger seats)         NaN      NaN     NaN
Motorcycle 125cc and under               NaN      NaN     NaN
Motorcycle 50cc and under                NaN      NaN     NaN
Motorcycle over 125cc and up to 500cc    NaN      NaN     NaN
Motorcycle over 500cc                    NaN      NaN     NaN
Other vehicle                            NaN      NaN     NaN
Pedal cycle                              NaN      NaN     NaN
Ridden horse                             NaN      NaN     NaN
Taxi/Private hire car                    NaN      NaN     NaN
Van / Goods 3.5 tonnes mgw or under      NaN      NaN     NaN

# INSIGTH # 13


In [46]:
df_exam_severity = df_exam3.groupby(['Vehicle_Type', 'Accident_Severity']).size().unstack()
df_exam_severity.unstack()

Accident_Severity  Vehicle_Type                         
Fatal              Agricultural vehicle                     0
                   Bus or coach (17 or more pass seats)     0
                   Car                                      0
                   Data missing or out of range             0
                   Goods 7.5 tonnes mgw and over            0
                   Goods over 3.5t. and under 7.5t          0
                   Minibus (8 - 16 passenger seats)         0
                   Motorcycle 125cc and under               0
                   Motorcycle 50cc and under                0
                   Motorcycle over 125cc and up to 500cc    0
                   Motorcycle over 500cc                    0
                   Other vehicle                            0
                   Pedal cycle                              0
                   Ridden horse                             0
                   Taxi/Private hire car                    0
             

# INSIGTH # 14

# In 2019, cars were the most involved vehicle type in accidents. However, over 80% of these accidents resulted in "Slight" injuries, with only a small percentage classified as "Fatal" or "Serious." This suggests that while cars have the highest accident count, most incidents are not life-threatening.



# INSIGTH # 15

# Despite having fewer total accidents than cars, motorcycles account for a disproportionately high number of "Fatal" and "Serious" accidents. Around 30% of motorcycle accidents resulted in serious injuries, and a significant portion were fatal, highlighting the vulnerability of motorcyclists on the road.

# INSIGTH # 16

#  Pedal cyclists and pedestrians involved in accidents have a much higher rate of "Serious" and "Fatal" injuries compared to other vehicle types. This suggests that road safety measures should prioritize protecting vulnerable road users.

In [47]:
accident.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  object        
 1   Accident_Severity        660679 non-null  category      
 2   Accident Date            660679 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  category      
 4   Light_Conditions         660679 non-null  category      
 5   District Area            660679 non-null  category      
 6   Longitude                660679 non-null  category      
 7   Number_of_Casualties     660679 non-null  category      
 8   Number_of_Vehicles       660679 non-null  category      
 9   Road_Surface_Conditions  660679 non-null  category      
 10  Road_Type                660679 non-null  category      
 11  Urban_or_Rural_Area      660679 non-null  category      
 12  Weather_Conditio

In [50]:
df_exam_road = df_exam3.groupby(['Road_Type', 'Accident_Severity']).size().unstack()
df_exam_road

Accident_Severity   Fatal  Serious  Slight
Road_Type                                 
Dual carriageway        0        0       0
One way street          0        0       0
Roundabout              0        0       0
Single carriageway      0        0       0
Slip road               0        0       0

# INSIGTHS # 17
# Most fatal accidents occur on high-speed roads like Motorways and Dual Carriageways, highlighting the impact of high-speed collisions

# INSIGTH # 18
# Urban roads (Single Carriageways) have the highest number of slight accidents due to frequent stops, intersections, and congestion.

# INSIGTH # 19
# Rural roads have a higher proportion of fatal accidents compared to urban roads, possibly due to fewer safety measures like barriers and lighting.

# INSIGTH # 20
# Roundabouts and Junctions show a high percentage of slight accidents, often due to misjudged turns or failing to yield.